### PR TITLE
clickhouse copier

### PR DESCRIPTION
**Pull request by nikitamikhaylov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Now clickhouse-copier splits each partition in number of pieces and copies them independently. 

Detailed description / Documentation draft:

...

By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.

### CODE REVIEW

**Comment by nvartolomei** [comment]
🥳

> Now clickhouse-copier splits each partition in number of pieces and copies them independently.
Why is this "Performance Improvement"? Because we can use multiple replicas as partition source or for other reason?

How atomicity is handled in this case? On error abort all jobs, drop partition and try again as before?

**Comment by nikitamikhaylov** [comment]
Earlier when Copier faced some troubles, it had to drop the whole partition.  Also each partition may be "smeared" to many shards (month as a partition key for example) because of using Distributed table. Operation of copying each partition was successful only if all the shards that contain particular partition done copying without crash of failure. And if you had large partition, the probability of crash was extremely high during copying. By splitting each partition you can increase the effectiveness (you won't drop huge amount of data). Atomicity is handled by ZooKeeper. In addition inserting will be not into destination table, but to auxiliary table with partition_key equals to old_partition_key % n or something like that.

**Comment by nikitamikhaylov** [comment]
How we split out partition:

**Comment by nikitamikhaylov** [comment]
@nvartolomei Do you use copier? Do you disable merges before copying?

**Comment by nvartolomei** [comment]
> By splitting each partition you can increase the effectiveness (you won't drop huge amount of data).
True, but I can't see any changes in this PR that would change this behaviour.

> Do you use copier? Do you disable merges before copying?
Used it a few times to split clusters. Didn't disable merges.

**Comment by nikitamikhaylov** [comment]
You can't see this changes, because I'm currently working on this task and don't publish commits.

**Comment by nikitamikhaylov** [comment]
And now you can't see significant changes because of "refactoring" and big diff. It was too difficult for me to understand huge single file. I will split this PR into two.

**Comment by nikitamikhaylov** [comment]
last commit is incorrect

File: dbms/programs/copier/ClusterCopier.cpp [hunk 1]
```diff
@@ -738,20 +980,50 @@ bool ClusterCopier::tryProcessTable(const ConnectionTimeouts & timeouts, TaskTab
...
+                try
+                {
+                    auto res = tryMoveAllPiecesToDestinationTable(task_table, partition_name);
+                    if (res == PartitionTaskStatus::Finished)
+                    {
+                        partition_moving_is_done = true;
+                        break;
+                    }
+
+                    /// Sleep if this task is active
```
Comment (nikitamikhaylov) [inline dbms/programs/copier/ClusterCopier.cpp]: Wrong comment.

**Comment by miggi** [comment]
@nikitamikhaylov @Akazz hi, already asked in Telegram channel, but they recommend me to ask here. New version of the ClickHouse Copier has started spawning series of tables which seems pretty weird: 

│ tbl_local_piece_1        │
│ tbl_local_piece_2        │
│ tbl_local_piece_3        │
│ tbl_local_piece_4        │
│ tbl_local_piece_5        │
│ tbl_local_piece_6        │
│ tbl_local_piece_7        │
│ tbl_local_piece_8        │

is this expected behavior ? how we should clean them after transfer ? manually or automatically after completion of migration ? 
As an idea - we might need to document the behaviour somewhere. 
Thanks for answer!

**Comment by nikitamikhaylov** [comment]
> @nikitamikhaylov @Akazz hi, already asked in Telegram channel, but they recommend me to ask here. New version of the ClickHouse Copier has started spawning series of tables which seems pretty weird: 
> 
> 
> 
> │ tbl_local_piece_1        │
> 
> │ tbl_local_piece_2        │
> 
> │ tbl_local_piece_3        │
> 
> │ tbl_local_piece_4        │
> 
> │ tbl_local_piece_5        │
> 
> │ tbl_local_piece_6        │
> 
> │ tbl_local_piece_7        │
> 
> │ tbl_local_piece_8        │
> 
> 
> 
> is this expected behavior ? how we should clean them after transfer ? manually or automatically after completion of migration ? 
> 
> As an idea - we might need to document the behaviour somewhere. 
> 
> Thanks for answer! 

Hello! These tables will be removed automatically after the copying is done. Also you can specify the number of pieces to split each partition. Is is needed to increase the fault tolerance due to networks failures.

Maybe, I will update our documentation later :)

**Comment by miggi** [comment]
Cool, which parameter I should use for that ? And probably subjective, but currently copier works significantly slower that "old one" insert from remote(..)

**Comment by miggi** [comment]
And.. unfortunately as I see the speed of import ~3x times slower that might be on one-node insert-from remote(..). I had to use non-copier version now to migrate data

**Comment by nikitamikhaylov** [comment]
> Cool, which parameter I should use for that ? And probably subjective, but currently copier works significantly slower that "old one" insert from remote(..)

`<number_of_splits>1</number_of_splits>` -- just add it to the copier's task.
How many copiers do you execute? There is no need to use only one process. 
And the new copier have to read partition more than once, so it is IO bounded.

**Comment by miggi** [comment]
Thanks,  I'm using one copier of course. 

Yes, but we had a similar option: 
    <max_workers>4</max_workers>
for that. As far I understood workers = number of import threads, ideally equal to number of shards.

**Comment by nikitamikhaylov** [comment]
> Thanks, I'm using one copier of course.
> 
> Yes, but we had a similar option:
> <max_workers>4</max_workers>
> for that. As far I understood workers = number of import threads, ideally equal to number of shards.

No, this setting is needed to control the number of copier processes for each task. A bit useless. 
For you to know, copier was designed to be executed on different machines in parallel. It uses Zookeeper to synchronize and to store metadata.

**Comment by miggi** [comment]
Alright thanks. I use default behaviour with number_of_splits = 10, but the clickhouse-copier seems to be too slow comparing to  previous version which I used a lot... and now I had to use now insert into with screen session.  

Also I've cleaned up all the tables manually and had to stop this.  

That's seems to me like downgrade :) you can consider this like a feedback guys.

**Comment by RO-29** [comment]
>Also I've cleaned up all the tables manually and had to stop this.

>That's seems to me like downgrade :) you can consider this like a feedback guys.

@miggi did you test with latest versions? I am using click house latest version v21.1.2 and still facing a slow copy issue

**Comment by miggi** [comment]
@RO-29 nope, new version of the clickhouse-copier seemed much more creepier to me :) I used copying select * from into with process under screen. It took some time) to move many Tb of data. But it's predictable and reliable. 

Let's hope for fixes.

**Comment by RO-29** [comment]
> select * from into
For rebalancing the shards, let's say I have to move 10Tb, approx how many hours will it take?

Assuming All my shards and cluster are on the same network (5GiB/s)

**Comment by miggi** [comment]
sorry, I meant insert into table A select * from table A . I used it for moving data of course. It took 5-8 days to move ~12 Tb (~150 billons of records). But its's very dependant on your bandwidth, cpu and other factors. I had different architecture on destination cluster, so I used insertion into distributed table (for me it was ~400k rows per second)

 With this approach stability is good - but I had few failures with "too long running query" i had to change it in CH config to some large value after that.  Please change max query duration - by default as far I remember 20 min. I changed it to 2 weeks.

**Pull request merged**